google-cloud-firestore/src/main/java/com/google/cloud/firestore/Order.java

```diff
@@ -136,9 +136,31 @@ public int compare(@Nonnull Value left, @Nonnull Value right) {
 
   /** Compare strings in UTF-8 encoded byte order */
   public static int compareUtf8Strings(String left, String right) {
-    ByteString leftBytes = ByteString.copyFromUtf8(left);
-    ByteString rightBytes = ByteString.copyFromUtf8(right);
-    return compareByteStrings(leftBytes, rightBytes);
+    int i = 0;
+    while (i < left.length() && i < right.length()) {
+      int leftCodePoint = left.codePointAt(i);
+      int rightCodePoint = right.codePointAt(i);
+
+      if (leftCodePoint != rightCodePoint) {
+        if (leftCodePoint < 128 && rightCodePoint < 128) {
+          // ASCII comparison
+          return Integer.compare(leftCodePoint, rightCodePoint);
+        } else {
+          // UTF-8 encoded byte comparison, substring 2 indexes to cover surrogate pairs
+          ByteString leftBytes =
+              ByteString.copyFromUtf8(left.substring(i, Math.min(i + 2, left.length())));
```

Comment:
Should this be `left.length() - i`?

Reply:
Nit: @milaGGL I would improve the following comment to be more descriptive (along the lines of what you said above):

changing
```java
// UTF-8 encoded byte comparison, substring 2 indexes to cover surrogate pairs
```
to
```java
// We have identified a code point difference, so we don't need to
// encode/compare _all_ the remainder of the strings. We only need to 
// compare up to 2 more indexes to cover potential surrogate pairs.
```